### PR TITLE
Column Schema protobuf and properties support

### PR DESCRIPTION
**Pull request by jperez999**
This PR provides support for writing ColumnSchemaSets to protobuf. It also updates the ColumnSchema to handle properties as a Dictionary (property_name, property_value).

### CODE REVIEW

File: nvtabular/columns/schema.py [hunk 1]
```diff
@@ -14,7 +14,96 @@
...
+from google.protobuf import json_format, text_format
+from google.protobuf.any_pb2 import Any
+from google.protobuf.struct_pb2 import Struct
+from tensorflow_metadata.proto.v0 import schema_pb2
+import numpy
+
+
+def register_extra_metadata(column_schema, feature):
+    msg_struct = Struct()
+    # msg_struct.update(col_schema.properties)
```
Comment (karlhigley) [inline nvtabular/columns/schema.py:29]:
```suggestion
```

File: nvtabular/columns/schema.py [hunk 2]
```diff
@@ -85,18 +202,87 @@ def select_by_name(self, names):
...
+            if len(feat.annotation.extra_metadata) > 1:
+                raise ValueError(
+                    f"{feat.name}: extra_metadata should have 1 item, has \
+                    {len(feat.annotation.extra_metadata)}"
+                )
+            if feat.annotation.extra_metadata:
+                properties = json_format.MessageToDict(feat.annotation.extra_metadata[0])["value"]
+            # what domain
+            # load the domain values
+            # import pdb; pdb.set_trace()
```
Comment (karlhigley) [inline nvtabular/columns/schema.py:233]:
```suggestion
```

File: nvtabular/columns/schema.py [hunk 3]
```diff
@@ -85,18 +202,87 @@ def select_by_name(self, names):
...
 
     def __add__(self, other):
         if other is None:
             return self
+        if not isinstance(other, Schema):
+            raise TypeError(
+                f"unsupported operand type(s) for +: 'Schema' and {type(other)}"
+            )
+        if other is None:
+            return self
```
Comment (karlhigley) [inline nvtabular/columns/schema.py:285]:
```suggestion
```
This check is already happening above

**Review by karlhigley — COMMENTED** [review]
Looks pretty good to me, left a few suggestions for cleaning up some stray code/comments

**Comment by karlhigley** [comment]
(It also needs some style fixes though)

File: nvtabular/columns/schema.py [hunk 4]
```diff
@@ -14,7 +14,101 @@
 # limitations under the License.
 #
 from dataclasses import dataclass, field
-from typing import List, Optional, Text
+from pathlib import Path
+from typing import Dict, List, Optional, Text
+
+import numpy
+from google.protobuf import json_format, text_format
```
Comment (benfred) [inline nvtabular/columns/schema.py:21]: We need to add something like 
https://github.com/NVIDIA/NVTabular/blob/e9c0795ab68d4d42b13d3a73e888c73b1483fb73/nvtabular/inference/triton/__init__.py#L25-L26

before any of these protobuf imports . I believe this is the root cause of the cpu CI failures you're seeing

Comment (jperez999) [reply nvtabular/columns/schema.py:21]: roger will fix now

File: nvtabular/columns/schema.py [hunk 5]
```diff
@@ -13,8 +13,106 @@
...
+            feature.value_count.CopyFrom(schema_pb2.ValueCount(min=min_length, max=max_length))
+        else:
+            # if no min max available set dummy value, to signal this is list
+            feature.value_count.CopyFrom(schema_pb2.ValueCount(min=0, max=0))
+    return feature
+
+
+def set_protobuf_float(column_schema, feature):
+    if str(column_schema.dtype) == ["float32", "float64"]:
+        # add emebeddings, tuple in properties (min, max)
```
Comment (benfred) [inline nvtabular/columns/schema.py:63]:
```suggestion
        # add embeddings, tuple in properties (min, max)
```

File: nvtabular/columns/schema.py [hunk 6]
```diff
@@ -13,8 +13,106 @@
...
+        col_min, col_max = column_schema.properties.pop("domain")
+        feature.float_domain.min = col_min
+        feature.float_domain.max = col_max
+        feature.type = 3
+    return feature
+
+
+def set_protobuf_int(column_schema, feature):
+    if str(column_schema.dtype) == ["int8", "int32", "int64"]:
+        # add emebeddings, tuple in properties (min, max)
```
Comment (benfred) [inline nvtabular/columns/schema.py:73]:
```suggestion
        # add embeddings, tuple in properties (min, max)
```

File: nvtabular/columns/schema.py [hunk 7]
```diff
@@ -13,8 +13,106 @@
...
+    return feature
+
+
+def set_protobuf_float(column_schema, feature):
+    if str(column_schema.dtype) == ["float32", "float64"]:
+        # add emebeddings, tuple in properties (min, max)
+        col_min, col_max = column_schema.properties.pop("domain")
+        feature.float_domain.min = col_min
+        feature.float_domain.max = col_max
+        feature.type = 3
```
Comment (benfred) [inline nvtabular/columns/schema.py:67]:
```suggestion
        feature.type = schema_pb2.FeatureType.FLOAT
```

File: nvtabular/columns/schema.py [hunk 8]
```diff
@@ -13,8 +13,106 @@
...
+    return feature
+
+
+def set_protobuf_int(column_schema, feature):
+    if str(column_schema.dtype) == ["int8", "int32", "int64"]:
+        # add emebeddings, tuple in properties (min, max)
+        col_min, col_max = column_schema.properties.pop("domain")
+        feature.int_domain.min = col_min
+        feature.int_domain.max = col_max
+        feature.type = 2
```
Comment (benfred) [inline nvtabular/columns/schema.py:77]:
```suggestion
        feature.type = schema_pb2.FeatureType.INT
```

File: nvtabular/columns/schema.py [hunk 9]
```diff
@@ -36,7 +137,32 @@ def with_tags(self, tags) -> "ColumnSchema":
...
+        return ColumnSchema(self.name, tags=self.tags, properties=properties)
+
+    def __eq__(self, other):
+        if not isinstance(other, ColumnSchema):
+            return False
+        if (
+            self.name == other.name
+            and self.tags == other.tags
+            and len(other.properties) == len(self.properties)
+            and self._is_list == other._is_list
```
Comment (benfred) [inline nvtabular/columns/schema.py:158]: Should we be using the dtype here too ?  If so, can we just use the default ```__eq__``` method provided by the dataclass ?

File: nvtabular/columns/schema.py [hunk 10]
```diff
@@ -13,8 +13,106 @@
...
+        # add emebeddings, tuple in properties (min, max)
+        col_min, col_max = column_schema.properties.pop("domain")
+        feature.float_domain.min = col_min
+        feature.float_domain.max = col_max
+        feature.type = 3
+    return feature
+
+
+def set_protobuf_int(column_schema, feature):
+    if str(column_schema.dtype) == ["int8", "int32", "int64"]:
```
Comment (benfred) [inline nvtabular/columns/schema.py:72]: Does this work with unsigned types?

File: nvtabular/columns/schema.py [hunk 11]
```diff
@@ -13,8 +13,106 @@
...
+        feature = proto_dict[string_name](column_schema, feature)
+    return feature
+
+
+proto_dict = {
+    "list": register_list,
+    "float": set_protobuf_float,
+    "int": set_protobuf_int,
+    "uint": set_protobuf_int,
+    "properties": register_extra_metadata,
```
Comment (benfred) [inline nvtabular/columns/schema.py:100]: is this 'properties' item here used?

File: nvtabular/columns/schema.py [hunk 12]
```diff
@@ -36,7 +137,32 @@ def with_tags(self, tags) -> "ColumnSchema":
 
         tags = list(set(list(self.tags) + tags))
 
-        return ColumnSchema(self.name, tags=tags)
+        return ColumnSchema(self.name, tags=tags, properties=self.properties)
```
Comment (benfred) [inline nvtabular/columns/schema.py:140]: I think this loses some information (dtype/is_list). how about: 

```suggestion
        return ColumnSchema(self.name, tags=tags, properties=self.properties, dtype=self.dtype, _is_list=self._is_list)
```

File: nvtabular/columns/schema.py [hunk 13]
```diff
@@ -36,7 +137,32 @@ def with_tags(self, tags) -> "ColumnSchema":
...
+
+    def with_properties(self, properties):
+        if not isinstance(properties, dict):
+            raise TypeError("properties must be in dict format, key: value")
+
+        # Using new dictionary to avoid passing old ref to new schema
+        properties.update(self.properties)
+
+        return ColumnSchema(self.name, tags=self.tags, properties=properties)
```
Comment (benfred) [inline nvtabular/columns/schema.py:149]:
```suggestion
        return ColumnSchema(self.name, tags=self.tags, properties=properties, dtype=self.dtype, _is_list=self._is_list)
```

File: nvtabular/columns/schema.py [hunk 14]
```diff
@@ -36,7 +137,32 @@ def with_tags(self, tags) -> "ColumnSchema":
...
+            self.name == other.name
+            and self.tags == other.tags
+            and len(other.properties) == len(self.properties)
+            and self._is_list == other._is_list
+        ):
+            # iterate through to ensure ALL keys AND values equal
+            for prop_name, prop_value in other.properties.items():
+                if prop_name not in self.properties or self.properties[prop_name] != prop_value:
+                    return False
+            return True
```
Comment (benfred) [inline nvtabular/columns/schema.py:164]: You can compare dictionaries with ==
```suggestion
            return self.properties == other.properties
```

File: nvtabular/columns/schema.py [hunk 15]
```diff
@@ -85,19 +211,82 @@ def select_by_name(self, names):
...
+            shape_name = feat.WhichOneof("shape_type")
+            if shape_name:
+                _is_list = True
+            field_name = feat.WhichOneof("domain_info")
+            if field_name:
+                domain_values = getattr(feat, field_name)
+                min_max = domain_values.min, domain_values.max
+                properties["domain"] = min_max
+            columns.append(
+                ColumnSchema(feat.name, tags=tags, properties=properties, _is_list=_is_list)
```
Comment (benfred) [inline nvtabular/columns/schema.py:251]: Can we get dtype information here too?

File: nvtabular/columns/schema.py [hunk 16]
```diff
@@ -85,19 +211,82 @@ def select_by_name(self, names):
...
+        return str([col_schema.__dict__ for col_schema in self.column_schemas.values()])
+
+    def __eq__(self, other):
+        if not isinstance(other, Schema) or len(self.column_schemas) != len(other.column_schemas):
+            return False
+        for col_name, col_schema in self.column_schemas.items():
+            # if not in or if not the same, Fail
+            if col_name not in other.column_schemas or other.column_schemas[col_name] != col_schema:
+                return False
+        return True
```
Comment (benfred) [inline nvtabular/columns/schema.py:283]:
```suggestion
        return self.column_schemas == other.column_schemas:
```

**Review by benfred — APPROVED** [review]

**Pull request merged**